### PR TITLE
Make sure locale doesn't change after flush is called

### LINKED ISSUE #145: AutoRouteManager doesn't re-translate objects back to current translation
Hi,
The AutoRouteManager translates Documents onFlush to any available locale in order to generate their Uri:

`Line 101: $this->adapter->translateObject($uriContextCollection->getSubjectObject(), $locale);`

But does not translate them back to the translation they had right before they were persisted/flushed.
In my case this leads to that I need to translate my object back:

`$document = $dm->findTranslation(null, '/cms/document', 'de');`
`// => available locales: en, de`
`$document->setTitle('New title');`

`$dm = $this->get('doctrine_phpcr.odm.document_manager');`
`$dm->persist($page);`
`$dm->flush();`

`//  $document->getLocale() => 'en'`

Is this intended, or am I missing something? ;-)

Cheers

### PR DESCRIPTION
**Pull request by WouterJ**
Fixes #145

I could not find an easy way to test this...

### CODE REVIEW

File: Doctrine/Phpcr/AutoRouteListener.php [hunk 1]
```diff
@@ -73,6 +75,14 @@ public function onFlush(ManagerEventArgs $args)
                     $dm->persist($autoRoute);
                     $uow->computeChangeSets();
                 }
+
+                // reset locale to the original locale
+                if (null !== $locale) {
+                    $meta = $dm->getMetadataFactory()->getMetadataFor(get_class($document));
+                    $dm->findTranslation($meta->getName(), $meta->getIdentifierValue($document), $locale);
```
Comment (dbu) [inline Doctrine/Phpcr/AutoRouteListener.php]: you should use $uow->getDocumentId($document) instead. reading the $meta->getIdentifierValue code, it seems that the id could be not mapped on the document.

File: Doctrine/Phpcr/AutoRouteListener.php [hunk 2]
```diff
@@ -73,6 +75,14 @@ public function onFlush(ManagerEventArgs $args)
...
+
+                    $locale = null;
```
Comment (dbu) [inline Doctrine/Phpcr/AutoRouteListener.php]: this is not needed, locale is overwritten at the beginning of the next loop

**Comment by dantleech** [comment]
+1 nice work @WouterJ -- whats the problem with testing? too much to setup?

File: Doctrine/Phpcr/AutoRouteListener.php (hunk 1 above)
Comment (WouterJ) [inline Doctrine/Phpcr/AutoRouteListener.php]: Thanks, didn't know such a method existed on the uow.

**Comment by WouterJ** [comment]
Added a test & fixed the comments.

@dantleech couldn't think of a way to bind different locales and then make sure the locale is not the last bound locale before calling flush. Now I realized it's just as simple as calling findTranslation before flushing...

File: Doctrine/Phpcr/AutoRouteListener.php (hunk 1 above)
Comment (dbu) [inline Doctrine/Phpcr/AutoRouteListener.php]: as with the method to determine the locale, we where pondering to expose
it in the document manager as well, to make it more "official"

File: Doctrine/Phpcr/AutoRouteListener.php [hunk 3]
```diff
@@ -73,6 +74,13 @@ public function onFlush(ManagerEventArgs $args)
                     $dm->persist($autoRoute);
                     $uow->computeChangeSets();
                 }
+
+                // reset locale to the original locale
+                if (null !== $locale) {
+                    $dm->findTranslation(get_class($document), $uow->getDocumentId($document), $locale);
+
+                    $locale = null;
```
Comment (dbu) [inline Doctrine/Phpcr/AutoRouteListener.php]: this does nothing, its overwritten in the next iteration.

File: Doctrine/Phpcr/AutoRouteListener.php [hunk 4]
```diff
@@ -73,6 +74,13 @@ public function onFlush(ManagerEventArgs $args)
                     $dm->persist($autoRoute);
                     $uow->computeChangeSets();
                 }
+
+                // reset locale to the original locale
+                if (null !== $locale) {
```
Comment (dbu) [inline Doctrine/Phpcr/AutoRouteListener.php]: you could compare locale with $uow->getCurrentLocale to check if its needed to reset the translation

**Comment by dbu** [comment]
not sure what exactly causes the test failure - looks unrelated to this PR. the php class is `Nodename` not `NodeName` - is that case sensitive?

File: Doctrine/Phpcr/AutoRouteListener.php (hunk 4 above)
Comment (WouterJ) [inline Doctrine/Phpcr/AutoRouteListener.php]: As lukas mentioned `findTranslation` was caching the items, I was a bit unsure what would be better, performance-wise.

**Comment by WouterJ** [comment]
Updated

**Comment by dantleech** [comment]
Tests failing, unidentified index `path`

**Comment by WouterJ** [comment]
@dantleech this seems to be a bug in Doctrine PHPCR, when calling `findTranslation` before `flush` (it works before 1.2.0-RC4 btw).

This PR works perfectly, but the test doesn't. Should be skip it until the bug has been fixed, so we can merge this?

**Comment by dantleech** [comment]
It depends if we are going to release CMF before PHPCR-ODM 1.3. If we have found a regression it must be fixed before the 1.3 release. If we plan to release 1.3 before or at the same time as the next CMF round of releases, I would prefer to wait.

Could you create an issue on phpcr-odm? If nobody else looks I will try and tackle it on Saturday

**Comment by dbu** [comment]
phpcr-odm 1.3 is at RC2, i would hope we can release that very soon and
before the cmf release.

**Comment by WouterJ** [comment]
Build passes now, thanks @dantleech

**Comment by dbu** [comment]
so, should we merge this? or is https://github.com/doctrine/phpcr-odm/pull/665 still a problem?

**Comment by WouterJ** [comment]
No, @dantleech worked around that bug in the tests.

**Comment by dantleech** [comment]
Cheers @WouterJ

**Pull request merged**